File: src/core/boys.ts
/** Boys function — port of GANSU boys.hpp */

import { BOYS_GRID } from '../data/boysGrid';
import { A_TR, B_TR, LUT_XI_INTERVAL, LUT_NUM_XI, LUT_K_MAX } from './constants';

const A_RS = 0.064048916778075;
const B_RS = 28.487431543672;
// Maclaurin expansion parameters (reserved for future use)
// const AA = 0.03768724;
// const BB = 0.60549623;
// const CC = 6.32743473;
// const DD = 10.350421;

/** Approximate erf(x) via Abramowitz-Stegun 7.1.26 (max error ~1.5e-7) */
function erf(x: number): number {
  const sign = x < 0 ? -1 : 1;
  const t = 1.0 / (1.0 + 0.3275911 * Math.abs(x));
  const y = 1.0 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) * t + 0.254829592) * t * Math.exp(-x * x);
  return sign * y;
}

/** Single Boys function F_n(x) */
export function boysFunction(n: number, x: number): number {
  if (x === 0.0) {
    return 1.0 / (2 * n + 1);
  }
  if (x < A_TR * n + B_TR) {
    return taylorInterpolation(n, x);
  }
  return recurrenceSemiInfinite(n, x);
}

/** Range-separated Boys function for erf(ωr)/r (long-range) Coulomb operator.
 *  F_n^{erf}(x, η, ω) = ξ^(n+1/2) · F_n(ξ x),  where ξ = ω²/(η + ω²).
 *  Reduces to standard F_n when ω → ∞, and to 0 when ω → 0.
 *  @param n  highest order requested (returns F_0 ... F_n)
 *  @param x  standard Boys argument (η · |R_PQ|²)
 *  @param eta η = pq/(p+q) for the shell quartet
 *  @param omega range-separation parameter (Bohr⁻¹) */
export function boysAllOmega(n: number, x: number, eta: number, omega: number): Float64Array {
  if (omega <= 0) return new Float64Array(n + 1); // erf with ω=0 is zero
  const omega2 = omega * omega;
  const xi = omega2 / (eta + omega2);
  const xPrime = xi * x;
  const fStandard = boysAll(n, xPrime);
  const result = new Float64Array(n + 1);
  let factor = Math.sqrt(xi); // ξ^(0+1/2)
  for (let j = 0; j <= n; j++) {
    result[j] = factor * fStandard[j];
    factor *= xi; // ξ^(j+1+1/2) for next iteration
  }
  return result;
}

/** Compute Boys functions F_0(x), F_1(x), ..., F_n(x) */
export function boysAll(n: number, x: number): Float64Array {
  const boys = new Float64Array(n + 1);

  if (x === 0.0) {
    for (let j = 0; j <= n; j++) boys[j] = 1.0 / (2 * j + 1);
    return boys;
  }

  if (x < A_TR * n + B_TR) {
    // Taylor interpolation for all orders
    taylorInterpolationAll(n, x, boys);
  } else {
    // Recurrence from semi-infinite
    recurrenceSemiInfiniteAll(n, x, boys);
  }

  return boys;
}

function taylorInterpolation(n: number, x: number): number {
  const xIdx = Math.min(Math.round(x / LUT_XI_INTERVAL), LUT_NUM_XI - 1);
  const deltaX = x - LUT_XI_INTERVAL * xIdx;
  let numerator = 1.0;
  let factorial = 1;
  let Fx = BOYS_GRID[LUT_NUM_XI * n + xIdx];

  for (let k = 1; k <= LUT_K_MAX; k++) {
    numerator *= -deltaX;
    factorial *= k;
    Fx += (BOYS_GRID[LUT_NUM_XI * (n + k) + xIdx] * numerator) / factorial;
  }
  return Fx;
}

function taylorInterpolationAll(n: number, x: number, boys: Float64Array) {
  const xIdx = Math.min(Math.round(x / LUT_XI_INTERVAL), LUT_NUM_XI - 1);
  const deltaX = x - LUT_XI_INTERVAL * xIdx;

  for (let j = 0; j <= n; j++) {
    let numerator = 1.0;
    let factorial = 1;
    let Fx = BOYS_GRID[LUT_NUM_XI * j + xIdx];

    for (let k = 1; k <= LUT_K_MAX; k++) {
      numerator *= -deltaX;
      factorial *= k;
      Fx += (BOYS_GRID[LUT_NUM_XI * (j + k) + xIdx] * numerator) / factorial;
    }
    boys[j] = Fx;
  }
}

function recurrenceSemiInfinite(n: number, x: number): number {
  let expNegX = 0.0;
  const recip2x = 1.0 / (2.0 * x);
  let Fx = 0.5 * Math.sqrt(Math.PI / x);

  if (x < A_RS * n + B_RS) {
    expNegX = Math.exp(-x);
    Fx *= erf(Math.sqrt(x));
  }

  for (let j = 1; j <= n; j++) {
    Fx = ((2 * j - 1) * Fx - expNegX) * recip2x;
  }
  return Fx;
}

function recurrenceSemiInfiniteAll(n: number, x: number, boys: Float64Array) {
  let expNegX = 0.0;
  const recip2x = 1.0 / (2.0 * x);
  let Fx = 0.5 * Math.sqrt(Math.PI / x);

  if (x < A_RS * n + B_RS) {
    expNegX = Math.exp(-x);
    Fx *= erf(Math.sqrt(x));
  }

  boys[0] = Fx;
  for (let j = 1; j <= n; j++) {
    Fx = ((2 * j - 1) * Fx - expNegX) * recip2x;
    boys[j] = Fx;
  }
}

File: wasm-eri/src/boys.rs
/// Boys function — port of GANSU boys.hpp / boys.ts

use crate::boys_grid::BOYS_GRID;

const A_TR: f64 = 0.352905920120321;
const B_TR: f64 = 0.015532762923351;
const A_RS: f64 = 0.064048916778075;
const B_RS: f64 = 28.487431543672;
const LUT_XI_INTERVAL: f64 = 0.03125;
const LUT_NUM_XI: usize = 1024;
const LUT_K_MAX: usize = 5;
const LUT_N_RANGE: usize = 24;

/// Approximate erf(x) via Abramowitz-Stegun 7.1.26
fn erf_approx(x: f64) -> f64 {
    let sign = if x < 0.0 { -1.0 } else { 1.0 };
    let t = 1.0 / (1.0 + 0.3275911 * x.abs());
    let y = 1.0
        - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) * t
            + 0.254829592)
            * t
            * (-x * x).exp();
    sign * y
}

/// Compute Boys functions F_0(x) through F_n(x) into `boys` slice.
pub fn boys_all(n: usize, x: f64, boys: &mut [f64]) {
    if x == 0.0 {
        for j in 0..=n {
            boys[j] = 1.0 / (2 * j + 1) as f64;
        }
        return;
    }

    if x < A_TR * n as f64 + B_TR {
        taylor_interpolation_all(n, x, boys);
    } else {
        recurrence_semi_infinite_all(n, x, boys);
    }
}

fn taylor_interpolation_all(n: usize, x: f64, boys: &mut [f64]) {
    let x_idx = ((x / LUT_XI_INTERVAL).round() as usize).min(LUT_NUM_XI - 1);
    let delta_x = x - LUT_XI_INTERVAL * x_idx as f64;

    for j in 0..=n {
        let mut numerator = 1.0_f64;
        let mut factorial = 1_u64;
        let mut fx = BOYS_GRID[LUT_NUM_XI * j + x_idx];

        let k_max = LUT_K_MAX.min(LUT_N_RANGE - j);
        for k in 1..=k_max {
            numerator *= -delta_x;
            factorial *= k as u64;
            fx += BOYS_GRID[LUT_NUM_XI * (j + k) + x_idx] * numerator / factorial as f64;
        }
        boys[j] = fx;
    }
}

fn recurrence_semi_infinite_all(n: usize, x: f64, boys: &mut [f64]) {
    let mut exp_neg_x = 0.0_f64;
    let recip2x = 1.0 / (2.0 * x);
    let mut fx = 0.5 * (std::f64::consts::PI / x).sqrt();

    if x < A_RS * n as f64 + B_RS {
        exp_neg_x = (-x).exp();
        fx *= erf_approx(x.sqrt());
    }

    boys[0] = fx;
    for j in 1..=n {
        fx = ((2 * j - 1) as f64 * fx - exp_neg_x) * recip2x;
        boys[j] = fx;
    }
}
